src/triage/component/catwalk/storage.py

```diff
@@ -339,30 +340,101 @@ def __init__(
+        indices = self.metadata['indices']
+        if matrix_with_labels.index.names != indices:
+            matrix_with_labels.set_index(indices, inplace=True)
+        matrix_with_labels = downcast_matrix(matrix_with_labels)
+        return matrix_with_labels
+
+    def _split_matrix(self, matrix_with_labels):
+        """Split a matrix with labels into a design matrix and label vector
+
+        Returns both as a tuple
+        """
+        labels = matrix_with_labels.pop(self.label_column_name)
+        # After we pop the labels from the matrix with labels it is now a design matrix
+        design_matrix = matrix_with_labels
+        if self.should_cache:
+            # we neither want to store multiple copies of the matrix
+            # nor misrepresent the label-less version of the matrix as the one with labels
+            # so we explicitly set the matrix_with_labels to None in the cache
+            self.matrix_with_labels = None
+            self.design_matrix = design_matrix
+            self.labels = labels
+        return design_matrix, labels
 
     @property
-    def matrix(self):
-        """The raw matrix. Will load from storage into memory if not already loaded"""
-        if self.__matrix is None:
-            self.__matrix = self._load()
-            # Is the index already in place?
-            if self.__matrix.index.names != self.metadata['indices']:
-                self.__matrix.set_index(self.metadata['indices'], inplace=True)
+    def design_matrix(self):
+        """The matrix without the label vector, only the index and features"""
+        if self.__design_matrix is not None:
+            return self.__design_matrix
+        else:
+            return self._split_matrix(self.matrix_with_labels)[0]
 
-            self.__matrix = downcast_matrix(self.__matrix)
-        return self.__matrix
+    @design_matrix.setter
+    def design_matrix(self, design_matrix):
+        self.__design_matrix = design_matrix
 
-    @matrix.setter
-    def matrix(self, matrix):
-        self.__matrix = matrix
+    @property
+    def labels(self):
+        """The label vector of the matrix"""
+        if self.__labels is not None:
+            return self.__labels
+        return self._split_matrix(self.matrix_with_labels)[1]
+
+    @labels.setter
+    def labels(self, labels):
+        self.__labels = labels
+
+    @property
+    def matrix_with_labels(self):
+        """The matrix complete with label vector."""
+        if self.__matrix_with_labels is not None:
+            return self._preprocess_matrix_with_labels(self.__matrix_with_labels)
+        elif self.__design_matrix is not None and self.__labels is not None:
+            return self.design_matrix.assign(**{self.label_column_name: self.labels})
+        else:
+            return self._preprocess_matrix_with_labels(self._load())
+
+    @matrix_with_labels.setter
+    def matrix_with_labels(self, matrix_with_labels):
+        self.__matrix_with_labels = matrix_with_labels
```

Comment:
I know it sounds evil popping the labels from the original instance but we don't want to store these matrices twice in memory. Copying before popping is a no-go. 

Users want either the design matrix, or the labels, or both separately. They really never want the matrix with labels.  We *store* the matrix with labels, and it's easiest to construct them that way in tests and such, but otherwise we don't want to push that interface. 

I also think the scheme I've come up with is kind of complicated, and likely unnecessarily so.  I don't necessarily mind making the main interface into the split_matrix tuple if that helps simplify things, and another route would be to try push the matrix-with-labels further away from interface and just be a part of the storage access layer, never being cached no matter what (only design_matrix and labels are cached). Doing this would require more tests to be changed but could be worth it.

Reply:
Perhaps this doesn't apply; but, say I have a big heavy dictionary:

    big_data = {'a': BIGA, 'b': BIGB}

But I don't want it like that. I want just `BIGA` and just `BIGB`:

    biga = big_data['a']
    remainder = {'b': big_data['b']}

Or, identically:

    big_data2 = big_data.copy()
    biga = big_data2.pop('a')
    remainder = big_data2

In this case, I'm not storing my big data (matrices) twice in memory. It's a shallow copy, not a deep copy -- (and I would suggest it be that way, if for some reason it's not actually). I'm copying pointers, but so long as the underlying data (design_matrix and labels) isn't explicitly copied, the effect on memory should be completely negligible.

If the problem is that in fact the matrix/dict has many many keys, and so it's not kosher to do a shallow copy, then is it a problem to simply leave the dict as it is, and only .get() the labels?

It sounds like you're saying that that's not the case. But, for that matter, can it not eagerly split the data up, and only worry about that, and provide method(s) to supply them together as needed, (if that's ever really necessary)?

(I think that's basically your last suggestion.)

Regardless, I would hope the interface and design could be _simpler_ -- and that this could, and should, result in _simpler_ tests -- though, certainly, still plenty of tests, and I'm sure, additional work on the tests, (which I understand might not be desired!).

Reply:
In recent commits I split the data up more (I took my own last suggestion), take another look.